nomos-core/chain-defs/src/proofs/leader_proof.rs

```diff
@@ -1,100 +1,257 @@
+    #[serde(with = "serde_fr")]
+    pub latest_root: Fr,
+}
+
+impl LeaderPublic {
+    #[must_use]
+    pub const fn new(
+        aged_root: Fr,
+        latest_root: Fr,
+        epoch_nonce: Fr,
+        slot: u64,
+        total_stake: u64,
+    ) -> Self {
+        Self {
+            slot,
+            epoch_nonce,
+            total_stake,
+            aged_root,
+            latest_root,
+        }
     }
 
-    fn entropy(&self) -> [u8; 32] {
-        self.public_inputs.entropy
+    #[must_use]
+    pub fn check_winning(&self, value: u64, note_id: Fr, sk: Fr) -> bool {
+        let (t0, t1) = self.scaled_phi_approx();
+        let threshold =
+            Self::phi_approx(&Fr::from(value), &(Fr::from(t0), Fr::from(t1))).into_bigint();
+        let ticket = Self::ticket(note_id, sk, self.epoch_nonce, Fr::from(self.slot)).into_bigint();
+        ticket < threshold
+    }
+
+    fn scaled_phi_approx(&self) -> (BigUint, BigUint) {
+        let t0 = &*pol::T0_CONSTANT / &BigUint::from(self.total_stake);
+        let total_stake_sq = &BigUint::from(self.total_stake) * &BigUint::from(self.total_stake);
+        let t1 = &*pol::P - (&*pol::T1_CONSTANT / &total_stake_sq);
+        (t0, t1)
+    }
+
+    fn phi_approx(stake: &Fr, approx: &(Fr, Fr)) -> Fr {
+        // stake * (t0 - t1 * stake)
+        *stake * (approx.0 - (approx.1 * *stake))
+    }
+
+    fn ticket(note_id: Fr, sk: Fr, epoch_nonce: Fr, slot: Fr) -> Fr {
+        Poseidon2Bn254Hasher::digest(&[note_id, sk, epoch_nonce, slot])
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LeaderPrivate {
+    input: pol::PolWitnessInputs,
+    pk: ed25519_dalek::VerifyingKey,
+}
+
+impl LeaderPrivate {
+    #[must_use]
+    pub fn new(
+        public: LeaderPublic,
+        note: Utxo,
+        aged_path: &MerklePath<Fr>,
+        latest_path: &MerklePath<Fr>,
+        slot_secret: Fr,
+        starting_slot: u64,
+        leader_pk: &ed25519_dalek::VerifyingKey,
+    ) -> Self {
+        let public_key = *leader_pk;
+        let leader_pk = ed25519_pk_to_fr_tuple(leader_pk);
+        let chain = pol::PolChainInputsData {
+            slot_number: public.slot,
+            epoch_nonce: public.epoch_nonce,
+            total_stake: public.total_stake,
+            aged_root: public.aged_root,
+            latest_root: public.latest_root,
+            leader_pk,
+        };
+        let wallet = pol::PolWalletInputsData {
+            note_value: note.note.value,
+            transaction_hash: *note.tx_hash.as_ref(),
+            output_number: note.output_index as u64,
+            aged_path: aged_path.iter().map(|n| *n.item()).collect(),
+            aged_selector: aged_path
+                .iter()
+                .map(|n| matches!(n, MerkleNode::Right(_)))
+                .collect(),
+            latest_path: latest_path.iter().map(|n| *n.item()).collect(),
+            latest_selector: latest_path
+                .iter()
+                .map(|n| matches!(n, MerkleNode::Right(_)))
```

Comment:
I've asked in Notion to clarify how these selectors are created: https://www.notion.so/Proof-of-Leadership-Specification-21c261aa09df819ba5b6d95d0fe3066d?d=270261aa09df80c9990c001c789991e1&source=copy_link#21c261aa09df81d49e68ee9b67269508

Reply:
It's been updated now in the spec. It was missing before